From 1d1588f3d3a244d2e3dffed86af44a0f9ab944de Mon Sep 17 00:00:00 2001
From: David Subiros <david.perez5@hp.com>
Date: Wed, 28 Sep 2011 15:19:04 +0100
Subject: [PATCH] Enforce snapshot cleanup.

Makes sure that the snapshot and the temporary directory is cleaned up,
even if qemu-img fails. Fixes bug 861582.

(cherry picked from commit eb6432361ee4946d93867e2d3f0870d892a25c66)

Change-Id: I718021bfb7069c39a47e0da25d79acbf07f02a69
---
 Authors                         |    1 +
 nova/virt/libvirt/connection.py |   48 ++++++++++++++++++++------------------
 2 files changed, 26 insertions(+), 23 deletions(-)

diff --git a/Authors b/Authors
index f3962c8..f23565b 100644
--- a/Authors
+++ b/Authors
@@ -28,6 +28,7 @@ Dan Prince <dan.prince@rackspace.com>
 Dan Wendlandt <dan@nicira.com>
 Dave Walker <DaveWalker@ubuntu.com>
 David Pravec <David.Pravec@danix.org>
+David Subiros <david.perez5@hp.com>
 Dean Troyer <dtroyer@gmail.com>
 Devendra Modium <dmodium@isi.edu>
 Devin Carlen <devin.carlen@gmail.com>
diff --git a/nova/virt/libvirt/connection.py b/nova/virt/libvirt/connection.py
index 7baeb69..67f4434 100644
--- a/nova/virt/libvirt/connection.py
+++ b/nova/virt/libvirt/connection.py
@@ -452,29 +452,31 @@ class LibvirtConnection(driver.ComputeDriver):
 
         # Export the snapshot to a raw image
         temp_dir = tempfile.mkdtemp()
-        out_path = os.path.join(temp_dir, snapshot_name)
-        qemu_img_cmd = ('qemu-img',
-                        'convert',
-                        '-f',
-                        source_format,
-                        '-O',
-                        image_format,
-                        '-s',
-                        snapshot_name,
-                        disk_path,
-                        out_path)
-        utils.execute(*qemu_img_cmd)
-
-        # Upload that image to the image service
-        with open(out_path) as image_file:
-            image_service.update(context,
-                                 image_href,
-                                 metadata,
-                                 image_file)
-
-        # Clean up
-        shutil.rmtree(temp_dir)
-        snapshot_ptr.delete(0)
+        try:
+            out_path = os.path.join(temp_dir, snapshot_name)
+            qemu_img_cmd = ('qemu-img',
+                            'convert',
+                            '-f',
+                            source_format,
+                            '-O',
+                            image_format,
+                            '-s',
+                            snapshot_name,
+                            disk_path,
+                            out_path)
+            utils.execute(*qemu_img_cmd)
+
+            # Upload that image to the image service
+            with open(out_path) as image_file:
+                image_service.update(context,
+                                     image_href,
+                                     metadata,
+                                     image_file)
+
+        finally:
+            # Clean up
+            shutil.rmtree(temp_dir)
+            snapshot_ptr.delete(0)
 
     @exception.wrap_exception()
     def reboot(self, instance, network_info, xml=None):
-- 
1.7.6.5

